handler/rfc8628/auth_handler.go

```diff
@@ -55,3 +47,39 @@ func (d *DeviceAuthHandler) HandleDeviceEndpointRequest(ctx context.Context, dar
 	resp.SetInterval(int(d.Config.GetDeviceAuthTokenPollingInterval(ctx).Seconds()))
 	return nil
 }
+
+func (d *DeviceAuthHandler) handleDeviceCode(ctx context.Context, dar fosite.DeviceRequester) (string, error) {
+	code, signature, err := d.Strategy.GenerateDeviceCode(ctx)
+	if err != nil {
+		return "", errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	dar.GetSession().SetExpiresAt(fosite.DeviceCode, time.Now().UTC().Add(d.Config.GetDeviceAndUserCodeLifespan(ctx)))
+	if err = d.Storage.CreateDeviceCodeSession(ctx, signature, dar.Sanitize(nil)); err != nil {
+		return "", errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	return code, nil
+}
+
+func (d *DeviceAuthHandler) handleUserCode(ctx context.Context, dar fosite.DeviceRequester) (string, error) {
+	var err error
+	var userCode, signature string
+	maxAttempts := 3
+	// Retry when persisting user code fails
+	// Possible causes include database connection issue, constraints violations, etc.
+	for i := 0; i < maxAttempts; i++ {
+		userCode, signature, err = d.Strategy.GenerateUserCode(ctx)
+		if err != nil {
+			return "", err
+		}
+
+		dar.GetSession().SetExpiresAt(fosite.UserCode, time.Now().UTC().Add(d.Config.GetDeviceAndUserCodeLifespan(ctx)).Round(time.Second))
+		err = d.Storage.CreateUserCodeSession(ctx, signature, dar.Sanitize(nil))
+		if err == nil {
+			return userCode, nil
+		}
+	}
+
+	return "", errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
```

Comment:
```suggestion
        errMsg := fmt.Sprintf("exceeded user-code generation max attempts %v : %s", MAX_ATTEMPTS, err.Error())
	return "", errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
```

let's enhance the error with our own message, or add a logging on top, whatever works better for you